system/gsettings.py

```diff
@@ -0,0 +1,258 @@
+                'more_suggestions-ebay.scope', 'more_suggestions-ubuntushop.scope',
+                'more_suggestions-skimlinks.scope']
+      # Power button invokes shutdown.
+      - schema: org.gnome.settings-daemon.plugins.power
+        key: button-power
+        value: 'shutdown'
+      # Desktop colours and wallpaper.
+      - schema: org.gnome.desktop.background
+        key: primary-color
+        value: '#000000'
+      - schema: org.gnome.desktop.background
+        key: secondary-color
+        value: '#000000'
+      - schema: org.gnome.desktop.background
+        key: picture-uri
+        value: 'file:///usr/share/backgrounds/Beach_by_Renato_Giordanelli.jpg'
+'''
+
+# Standard python libraries
+import os
+import pwd
+import time
+import signal
+import subprocess
+
+try:
+    from gi.repository import Gio, GLib
+except ImportError:
+    HAS_GIO = False
+else:
+    HAS_GIO = True
+
+ENV_DBUS = 'DBUS_SESSION_BUS_ADDRESS'
+SESSION_MANAGERS = [
+    'gnome-session', 'mate-session', 'xfce4-session',
+    'cinnamon-session', 'icewm-session', 'openbox-session']
+
+def proc_name(pid):
+    """Return process name of process pid"""
+    try:
+        for attr in open("/proc/{0}/status".format(pid), 'r'):
+            if attr.startswith('Name:'):
+                return attr.split(':')[1].strip()
+        return None
+    except:
+        return None
+
+def proc_owner(pid):
+    """Return username of UID of process pid"""
+    for ln in open('/proc/{0}/status'.format(pid), 'r'):
+        if ln.startswith('Uid:'):
+            return int(ln.split()[1])
+
+def proc_environ(pid, varname=""):
+    """Return process environment variable of process pid"""
+    proc_env = open("/proc/{0}/environ".format(pid), 'r')
+    env = proc_env.read()
+    for env_var in env.split('\x00'):
+        if varname != "":
+            prefix = "{0}=".format(varname)
+            if env_var.startswith(prefix):
+                return env_var[len(prefix):]
+
+def bus_address(pid):
+    """Return the bus address for the specified process"""
+    return proc_environ(pid, ENV_DBUS)
+
+def new_session():
+    """Create a new dbus session"""
+    dbus_process = subprocess.Popen('dbus-launch', stdout=subprocess.PIPE)
+    for env_string in dbus_process.stdout.readlines():
+        env_var, env_value = env_string.split('=', 1)
+        os.environ[env_var] = env_value
+    return os.environ['DBUS_SESSION_BUS_PID'].replace('\n', '')
+
+
+class Gsettings(object):
+    def __init__(self):
+        # Determine the session pid.
+        self.session_pid = -1
+        self._cleanup = False
+
+        for procfile in os.listdir('/proc'):
+            if procfile.isdigit():
+                if proc_name(procfile) in SESSION_MANAGERS:
+                    self.session_pid = procfile
+
+        osuser = os.getuid()
+        if self.session_pid != -1:
+            sessionuser = proc_owner(self.session_pid)
```

Comment:
There is a race condition, if the PID is reused (or change between the time the PID is found and the time this file is opened again ).